internal/component/beyla/beyla_linux.go

```diff
@@ -0,0 +1,244 @@
+		IgnorePatterns: args.IgnorePatterns,
+		IgnoredEvents:  transform.IgnoreMode(args.IgnoredEvents),
+	}
+}
+
+func (args Attributes) Convert() beyla.Attributes {
+	return beyla.Attributes{
+		Kubernetes: transform.KubernetesDecorator{
+			Enable: transform.KubeEnableFlag(args.Kubernetes.Enable),
+		},
+	}
+}
+
+func (args Discovery) Convert() (services.DiscoveryConfig, error) {
+	srv, err := args.Services.Convert()
+	if err != nil {
+		return services.DiscoveryConfig{}, err
+	}
+	return services.DiscoveryConfig{
+		Services: srv,
+	}, nil
+}
+
+func (args Services) Convert() (services.DefinitionCriteria, error) {
+	var attrs services.DefinitionCriteria
+	for _, s := range args {
+		ports, err := stringToPortEnum(s.OpenPorts)
+		if err != nil {
+			return nil, err
+		}
+		paths, err := stringToRegexpAttr(s.Path)
+		if err != nil {
+			return nil, err
+		}
+		attrs = append(attrs, services.Attributes{
+			Name:      s.Name,
+			Namespace: s.Namespace,
+			OpenPorts: ports,
+			Path:      paths,
+		})
+	}
+	return attrs, nil
+}
+
+func New(opts component.Options, args Arguments) (*Component, error) {
+	reg := prometheus.NewRegistry()
+	c := &Component{
+		opts:   opts,
+		args:   args,
+		reload: make(chan struct{}, 1),
+		reg:    reg,
+	}
+
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	var cancel context.CancelFunc
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-c.reload:
+			// cancel any previously running exporter
+			if cancel != nil {
+				cancel()
+			}
+			newCtx, cancelFunc := context.WithCancel(ctx)
+			cancel = cancelFunc
+
+			c.mut.Lock()
+			cfg, err := c.args.Convert()
+			if err != nil {
+				return fmt.Errorf("failed to convert arguments: %w", err)
+			}
+			cfg.Prometheus.Registry = c.reg
+			c.mut.Unlock()
+			components.RunBeyla(newCtx, cfg)
+		}
+	}
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
+	c.mut.Lock()
+	baseTarget, err := c.baseTarget()
+	if err != nil {
+		return err
+	}
+	c.opts.OnStateChange(Exports{
+		Targets: []discovery.Target{baseTarget},
+	})
+	c.mut.Unlock()
+	select {
+	case c.reload <- struct{}{}:
+	default:
+	}
+	return nil
+}
+
+func (c *Component) baseTarget() (discovery.Target, error) {
+	data, err := c.opts.GetServiceData(http_service.ServiceName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get HTTP information: %w", err)
+	}
+	httpData := data.(http_service.Data)
+
+	return discovery.Target{
+		model.AddressLabel:     httpData.MemoryListenAddr,
+		model.SchemeLabel:      "http",
+		model.MetricsPathLabel: path.Join(httpData.HTTPPathForComponent(c.opts.ID), "metrics"),
+		"instance":             defaultInstance(),
+		"job":                  "beyla",
+	}, nil
+}
+
+func (c *Component) Handler() http.Handler {
+	return promhttp.HandlerFor(c.reg, promhttp.HandlerOpts{})
+}
```

Comment:
Can you help me understand what the `/metrics` endpoint will return metrics for? I read the docs but it wasn't immediately clear. 

Metrics specific to Beyla itself (and not any applications it is monitoring) should be registered to the registry given to the component. 

However, metrics specific to applications Beyla is monitoring should be done via this pattern, where it exposes those metrics separately and not over the default metrics endpoint.

Reply:
`/metrics` endpoint will return the RED metrics exposed by Beyla. I followed the prometheus' exporters flow component pattern to implement this, but if there's an easier way to do it, i'm happy to change it.

Reply:
But are they RED metrics _about_ Beyla? Or about applications Beyla is monitoring?

Reply:
The RED metrics of the configured applications for Beyla. I'm not exposing internal metrics for now.

Reply:
OK 👍 That was the thing I was confused about. If it's metrics for configuration applications, then it makes sense to follow the exporter pattern for those metrics. 

In the future when you get around to exposing internal metrics, those shouldn't be exposed here, and should rather be exposed via the registry passed to the `component.Options` to align with how debug metrics for other components work.

Reply:
Thanks for the explanation! Should I leave a comment or something?

Reply:
Couldn't hurt!